### PR TITLE
[extractor/bilibili] Support flac format.

### PR DESCRIPTION
**Pull request by jackyyf**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Bilibili newly added support for lossless FLAC support, which is
available at `window.__playinfo__.data.dash.flac.audio`.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/bilibili.py [hunk 1]
```diff
@@ -218,6 +218,8 @@ def _real_extract(self, url):
 
         durl = traverse_obj(video_info, ('dash', 'video'))
         audios = traverse_obj(video_info, ('dash', 'audio')) or []
+        if 'flac' in video_info['dash'] and video_info['dash']['flac']['audio'] is not None:
+            audios.append(video_info['dash']['flac']['audio'])
```
Comment (Lesmiscore) [inline yt_dlp/extractor/bilibili.py:222]:
```suggestion
        flac_audio = traverse_obj(video_info, ('dash', 'flac', 'audio'))
        if flac_audio:
            audios.append(flac_audio)
```
Always avoid indexing lists/dicts/tuples directly but do it via `traverse_obj`

Comment (jackyyf) [reply yt_dlp/extractor/bilibili.py]: Understood, working on it.

Comment (jackyyf) [reply yt_dlp/extractor/bilibili.py]: PTAL, this has been fixed according to suggestion.

**Comment by pukkandan** [comment]
Check if the format sorting works correctly with this. Add a test if possible (set `'params': {'format': 'ba'}`)

**Comment by jackyyf** [comment]
Sorry for left this PR idle for several days, I'm kind of busy in my daily work

> Check if the format sorting works correctly with this. Add a test if possible (set `'params': {'format': 'ba'}`)

Right now FLAC only works with Bilibili "BIG" account (aka Premium account), non-Premium account will result in `window.__playinfo__.data.dash.flac.audio == null`

Difference between Premium account and non-Premium account (Video link https://www.bilibili.com/video/BV1hG41147Wx):
`__playinfo__` for Premium account:
image

`__playinfo__` for non-Premium account:
image

I could workaround this with mocking of `__playinfo__`, but I could not find the best approach in the docs. I would be super grateful if you could point me to a correct path for mocking in yt-dlp :)

**Comment by pukkandan** [comment]
If account is needed, you don't need to add test

**Pull request merged**